### ISSUE TITLE
Feature: Worker service type

### ISSUE THREAD

**Issue by hnrc** [aws/copilot-cli#1246]
Copilot currently supports two service types,  `Load Balanced Web Services` and `Backend Service`. Both assumes and supports HTTP type services where a port is exposed either internally in the VPC or publicly to the Internet.

I'd be interested in a service type that do not expose any port at all. I'm thinking of a `Worker` service type for long running background services. A typical use case can be a service that crunches messages from a queue.

Something similar to [Heroku's Worker dynos](https://devcenter.heroku.com/articles/background-jobs-queueing).

Does this make sense?

**Comment by efekarakus** [comment]
Yup totally makes sense! In your situation how does the queue get populated? Are you subscribed to a SNS topic from a different microservice or something else?
Do you use SQS or EventBridge as a queue? Does it even matter 😛?

**Comment by hnrc** [comment]
> Are you subscribed to a SNS topic from a different microservice or something else?
> Do you use SQS or EventBridge as a queue? Does it even matter 😛?

Aye, the most typical setup looks a little like this:

```
+-------------+    +-------+    +-------+    +-------------+
|             |    |       |    |       |    |             |
|  Service A  +--->+  SNS  +--->+  SQS  +--->+  Service B  |
|             |    |       |    |       |    |             |
+-------------+    +-------+    +-------+    +-------------+
```

However, there are also use cases where the worker service feeds itself with work.

**Comment by mwarkentin** [comment]
+1 for this from us - in our existing apps we have several which use SQS, and a whole bunch more which consume tasks from RabbitMQ.

**Comment by mwarkentin** [comment]
Seems like this can be closed after #1478 was merged.

**Comment by efekarakus** [comment]
I think there might be still room for us to explore a nice UX around publishing and subscribing to events from services. 

My initial thought was to provide something like this in the manifests:
```yaml
# in serviceA's manifest
publish:
  topics:
  - name: "subscriptions"
```
```yaml
# in serviceB's manifest
subscribe:
  topics:
  - from: "serviceA"
    name: "subscriptions"
```

ServiceA could create an SNS topic named "serviceA-subscriptions" and ServiceB gets a SQS queue that gets subscribed to the same topic.
Or similarly ServiceA publishes to an EventBridge at the environment level (shared across services), and ServiceB's SQS queue becomes a target. 

This is very high-level, we haven't investigated the tradeoffs yet. At first glance, the first approach might result in cross stack imports which can be troublesome during deletions. But let us know if anyone has thoughts on what type of experience you'd like around pub/sub!

**Comment by efekarakus** [comment]
Hi folks! We have an official proposal for this service type here: #2550 we would love your feedback!

**Comment by efekarakus** [comment]
Hi folks! I forgot to close this issue but we have released the "Worker Service" pattern few days ago 🎉 
https://github.com/aws/copilot-cli/releases/tag/v1.10.0 

Documentation: 
* Overview: https://aws.github.io/copilot-cli/docs/concepts/services/#worker-service
* Code samples: https://aws.github.io/copilot-cli/docs/developing/publish-subscribe/
* Manifest: https://aws.github.io/copilot-cli/docs/manifest/worker-service/

**Comment by akshat-pyor** [comment]
can we call from service A to service A ??

> > Are you subscribed to a SNS topic from a different microservice or something else?
> > Do you use SQS or EventBridge as a queue? Does it even matter 😛?
> 
> Aye, the most typical setup looks a little like this:
> 
> ```
> +-------------+    +-------+    +-------+    +-------------+
> |             |    |       |    |       |    |             |
> |  Service A  +--->+  SNS  +--->+  SQS  +--->+  Service B  |
> |             |    |       |    |       |    |             |
> +-------------+    +-------+    +-------+    +-------------+
> ```
> 
> However, there are also use cases where the worker service feeds itself with work.

**Comment by huanjani** [comment]
Hello, @akshat-pyor!
Yes, a Worker Service type of workload may both `publish` and `subscribe`. See https://aws.github.io/copilot-cli/docs/manifest/worker-service/ for details on how to configure your manifest to do so.
Thanks!

**Comment by SoundsSerious** [comment]
@huanjani is there any plan to implement subscribe on backend services or load balanced web services. While less common I think there are still plenty of use cases where these systems need to respond to events

**Comment by Lou1415926** [comment]
@SoundsSerious Hello! Thank you for bringing this up. Short answer is this is not yet on the short-term plan. This is a new feature request, so feel free to submit a ["Feature request"](https://github.com/aws/copilot-cli/issues/new/choose) issue for this.

With that being said, this is likely achievable through addons or YAML/CDK overrides. Have you given those a try?

Edit: sorry, nvm about opening an issue - saw you already done that!

**Comment by SoundsSerious** [comment]
@Lou1415926 Great I've done that as of sunday https://github.com/aws/copilot-cli/issues/5826

To be honest with you I'm not really a fan of cloud formation / CDK so i would likely give my copilot instances permissions to subscribe / notify to anything they want to get around this.
